File: permutations.py
#
#  LeetCode 46. Permutations Python
#
class Solution(object):
    def permute(self, nums:list[int])->list[int]:
        result = []

        if len(nums) == 1:
            return [nums[:]]

        for i in range(len(nums)):
            n = nums.pop(0)
            perms = self.permute(nums)

            for perm in perms:
                perm.append(n)
            
            result.extend(perms)
            nums.append(n)
        return result

nums = [1,2,3]
sol = Solution()
ans = sol.permute(nums)
print(ans)

File: permutations.php
<?php
/**
 * 
 * LeetCode 46. Permutations PHP
 * 
 **/
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function permute($nums) {
        $result = [];
        if( count($nums) == 1 )
            return [$nums];

        foreach( range(0, count($nums)-1) as $pos ){
            $n = array_shift($nums);
            $perms = $this->permute($nums);

            foreach( $perms as $k => $perm ){
                $perms[$k] = array_merge( $perm, [$n]);
            }
            $result = array_merge( $result, $perms );
            $nums = array_merge( $nums, [$n] );
        }

        return $result;
    }
}

$nums = [1,2,3];
$sol = new Solution();
$ans = $sol->permute($nums);
print_r($ans);

